### PR TITLE
clean up trans_static_method_callee and friends

### PR DESCRIPTION
**Pull request by arielb1**
The old code was terribly ugly and was duplicated in several places.

r? @michaelwoerister

### CODE REVIEW

**Comment by michaelwoerister** [comment]
I'll take a look at it ASAP.

File: src/librustc_trans/trans/mir/did.rs [hunk 1]
```diff
@@ -116,34 +115,20 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
...
-                } = imp_substs.types.split();
-                let callee_substs = Substs::erased(
-                    subst::VecPerParamSpace::new(impl_type, impl_self, rcvr_method)
-                );
+                let callee_substs = impl_substs.with_method_from(substs);
                 let mth = tcx.get_impl_method(impl_def_id, callee_substs, mname);
-                let mthsubsts = tcx.mk_substs(mth.substs);
-                self.trans_fn_ref(bcx, ty, mthsubsts, mth.method.def_id)
+                let mth_substs = tcx.mk_substs(mth.substs);
+                self.trans_fn_ref(bcx, ty, mth_substs, mth.method.def_id)
```
Comment (nagisa) [inline src/librustc_trans/trans/mir/did.rs]: Lovely!

File: src/librustc_trans/trans/meth.rs [hunk 2]
```diff
@@ -216,33 +175,7 @@ pub fn trans_static_method_callee<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
...
-                types: impl_type,
-                selfs: impl_self,
-                fns: _
-            } = impl_substs.types.split();
-            let callee_substs =
-                Substs::erased(VecPerParamSpace::new(impl_type,
-                                                     impl_self,
-                                                     rcvr_method));
-
+            let callee_substs = impl_substs.with_method_from(&rcvr_substs);
```
Comment (michaelwoerister) [inline src/librustc_trans/trans/meth.rs]: Should we take care to erase regions here?

File: src/librustc_trans/trans/meth.rs [hunk 3]
```diff
@@ -256,6 +189,7 @@ pub fn trans_static_method_callee<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
                               idx)
         }
         _ => {
+            // FIXME
```
Comment (michaelwoerister) [inline src/librustc_trans/trans/meth.rs]: What's this about?

File: src/librustc_trans/trans/meth.rs [hunk 4]
```diff
@@ -285,11 +219,11 @@ fn trans_monomorphized_callee<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
...
             // those from the impl and those from the method:
-            let callee_substs =
-                combine_impl_and_methods_tps(
-                    bcx, MethodCallKey(method_call), vtable_impl.substs);
-
-            let mth = bcx.tcx().get_impl_method(impl_did, callee_substs, mname);
+            let meth_substs = node_id_substs(ccx,
+                                             MethodCallKey(method_call),
+                                             bcx.fcx.param_substs);
+            let impl_substs = vtable_impl.substs.with_method_from(&meth_substs);
```
Comment (michaelwoerister) [inline src/librustc_trans/trans/meth.rs]: Same as above: Before `combine_impl_and_methods_tps` would make sure we'd get something with erased regions, now we can't rely on it unless we know that the `Substs` in the vtable are already region-erased.

**Comment by michaelwoerister** [comment]
Very nice work! Much cleaner indeed.

Is there a specific reason you don't erase regions in some of the cases anymore? It probably doesn't matter most of the time, since the consumer of the `Substs` arguments will take care of erasing regions itself. A comment would be nice though.

r=me when you have taken care of that empty `FIXME`.

File: src/librustc_trans/trans/meth.rs (hunk 3 above)
Comment (arielb1) [inline src/librustc_trans/trans/meth.rs]: This causes ICEs

**Comment by arielb1** [comment]
fixed that

**Comment by arielb1** [comment]
We don't need to erase regions again because `fulfill_obligation` (in fact, `infer::drain_fulfillment_cx`) erases them for us.

**Pull request merged**